File: scripts/production/route_identity.py
"""Economic route identity, independent of exchange support and account holdings."""
from __future__ import annotations

import math
from typing import Any

from scripts.production.causal_performance import asset

FIELDS = (
    "route_type", "base_economic_asset", "candidate_asset",
    "candidate_trigger_active", "resolved_execution_asset", "resolution_reason",
    "signal_available_at",
)


def resolve_route(*, route_type: str, base_economic_asset: str,
                  candidate_asset: str, candidate_trigger_active: bool,
                  target_exposure: float, signal_available_at: str) -> dict[str, Any]:
    route = str(route_type).upper()
    base, candidate = asset(base_economic_asset), asset(candidate_asset)
    if not isinstance(candidate_trigger_active, bool):
        raise ValueError("Candidate trigger must be an explicit boolean")
    exposure = float(target_exposure)
    if not math.isfinite(exposure) or exposure < 0:
        raise ValueError("Invalid route exposure")
    if route == "CASH":
        if exposure != 0:
            raise ValueError("CASH route requires zero exposure")
        resolved, reason = "CASH", "cash_route"
    elif route == "BTC":
        resolved, reason = "BTC", "btc_route"
    elif route == "BASE":
        resolved, reason = base, "same_interval_base_holding"
        if candidate_trigger_active:
            raise ValueError("BASE cannot claim an active candidate route")
    elif route == "CANDIDATE":
        if not candidate_trigger_active:
            raise ValueError("CANDIDATE route requires its actual trigger")
        resolved, reason = candidate, "active_candidate_trigger"
    else:
        raise ValueError("Unknown route: " + route)
    if route != "CASH" and (resolved == "CASH" or exposure <= 0):
        raise ValueError("Risk route requires a concrete asset and positive exposure")
    return dict(route_type=route, base_economic_asset=base, candidate_asset=candidate,
                candidate_trigger_active=candidate_trigger_active,
                resolved_execution_asset=resolved, resolution_reason=reason,
                signal_available_at=signal_available_at, target_exposure=exposure)


def validate_target_identity(snapshot: dict[str, Any]) -> None:
    """Consumers cannot silently fall back to a weekly candidate or an old label."""
    missing = set(FIELDS) - set(snapshot)
    if missing:
        raise ValueError("Missing production route identity: " + ", ".join(sorted(missing)))
    intent = snapshot["execution_intent"]
    expected = resolve_route(**{k: snapshot[k] for k in FIELDS if k not in {
        "resolved_execution_asset", "resolution_reason"}},
        target_exposure=intent["target_exposure"])
    for key in ("resolved_execution_asset", "resolution_reason"):
        if snapshot[key] != expected[key]:
            raise ValueError("Production route derivation mismatch: " + key)
    for key in ("current_asset", "actual_held_asset", "authorized_tradable_asset"):
        if asset(snapshot[key]) != expected["resolved_execution_asset"]:
            raise ValueError("Production target identity mismatch: " + key)
    if intent["target_asset"] != expected["resolved_execution_asset"]:
        raise ValueError("Execution intent diverges from resolved production asset")
    if any(intent.get(k) != snapshot[k] for k in FIELDS):
        raise ValueError("Execution intent route lineage differs from Production Core")

File: web/src/server/multi-account-executor/route-identity.ts
/** Bound to the Python Production Core contract; never infer a target from a candidate. */
export const routeFields = ["route_type", "base_economic_asset", "candidate_asset", "candidate_trigger_active",
  "resolved_execution_asset", "resolution_reason", "signal_available_at"] as const;

export function validateResolvedRoute(production: Record<string, unknown>, intent: Record<string, unknown>): void {
  const embedded = production.execution_intent as Record<string, unknown> | undefined;
  if (!embedded || routeFields.some((key) => production[key] === undefined || embedded[key] !== production[key] || intent[key] !== production[key])) throw new Error("Canonical route lineage is missing or inconsistent.");
  const route = production.route_type;
  const active = production.candidate_trigger_active;
  if (typeof active !== "boolean") throw new Error("Candidate trigger must be explicit.");
  const expected = route === "CASH" ? "CASH" : route === "BTC" ? "BTC" : route === "BASE" && !active ? production.base_economic_asset : route === "CANDIDATE" && active ? production.candidate_asset : null;
  const reasons = { CASH: "cash_route", BTC: "btc_route", BASE: "same_interval_base_holding", CANDIDATE: "active_candidate_trigger" };
  if (typeof route !== "string" || !Object.hasOwn(reasons, route) || production.resolution_reason !== reasons[route as keyof typeof reasons]) throw new Error("Canonical route resolution reason is invalid.");
  if (typeof expected !== "string" || !/^[A-Z][A-Z0-9]*$/.test(expected) || ["BASE", "BASELINE", "CANDIDATE", "ALT"].includes(expected)) throw new Error("Concrete economic target is required.");
  if ([production.resolved_execution_asset, production.current_asset, production.actual_held_asset, embedded.target_asset, intent.target_asset].some((value) => value !== expected)) throw new Error("Canonical route target identity differs.");
  const exposure = embedded.target_exposure;
  if (typeof exposure !== "number" || !Number.isFinite(exposure) || (expected === "CASH" ? exposure !== 0 : exposure <= 0)) throw new Error("Canonical route exposure is invalid.");
  const available = Date.parse(String(production.signal_available_at));
  const close = Date.parse(`${production.closed_day}T00:00:00Z`) + 86_400_000;
  if (!Number.isFinite(available) || !Number.isFinite(close) || available <= close || available > Date.now()) throw new Error("Canonical signal availability is invalid.");
}
